### PR TITLE
Distinct metrics (inter- and intra-distinct)

### LINKED ISSUE #3255: Could eval_model compute the distinct metrics?
Hi, I would like to compute **the distinct metrics** using the `eval_model` command. 

But I saw the CLI Arguments (as follows), it likely could not support to compute **the distinct metrics**. Is it?

```
-mcs, --metrics | List of metrics to show/compute, e.g. all, default, or give a list split by, like ppl,f1, accuracy, hits@1, rouge, bleu. the rouge metrics will be computed as rouge-1, rouge-2, and rouge-l
```

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
Adds support for `interdistinct` and `intradistinct` metrics.

Fixes #3255 and replaces #3265.

**Testing steps**
New CI test for correctness. Manual eval_model for integration.

```
$ parlai em -t convai2 --model repeat_label --metrics distinct
18:21:08 | Opt:
18:21:08 |     aggregate_micro: False
18:21:08 |     allow_missing_init_opts: False
18:21:08 |     batchsize: 1
18:21:08 |     cant_answer_message: "I don't know."
18:21:08 |     cant_answer_percent: 0
18:21:08 |     datapath: /private/home/roller/working/parlai/data
18:21:08 |     datatype: valid
18:21:08 |     dict_class: None
18:21:08 |     display_examples: False
18:21:08 |     download_path: None
18:21:08 |     dynamic_batching: None
18:21:08 |     hide_labels: False
18:21:08 |     image_cropsize: 224
18:21:08 |     image_mode: raw
18:21:08 |     image_size: 256
18:21:08 |     init_model: None
18:21:08 |     init_opt: None
18:21:08 |     log_every_n_secs: 10
18:21:08 |     log_keep_fields: all
18:21:08 |     loglevel: info
18:21:08 |     metrics: distinct
18:21:08 |     model: repeat_label
18:21:08 |     model_file: None
18:21:08 |     multitask_weights: [1]
18:21:08 |     num_examples: -1
18:21:08 |     override: "{'task': 'convai2', 'model': 'repeat_label', 'metrics': 'distinct'}"
18:21:08 |     parlai_home: /private/home/roller/working/parlai
18:21:08 |     report_filename:
18:21:08 |     return_one_random_answer: True
18:21:08 |     save_format: conversations
18:21:08 |     save_world_logs: False
18:21:08 |     starttime: Jan05_18-21
18:21:08 |     task: convai2
18:21:08 |     tensorboard_log: False
18:21:08 |     tensorboard_logdir: None
18:21:08 |     verbose: False
18:21:08 | Current ParlAI commit: 31eb9ab6a7044d0300f678e36534ace456596d3e
18:21:08 | Current internal commit: c0abb1b7f5e6841c9d046c04e988744756dde675
18:21:08 | Current fb commit: 4db2d5690edc7b1fd5523a495de346dcbcb4bb57
18:21:08 | Evaluating task convai2 using datatype valid.
18:21:08 | creating task(s): convai2
18:21:08 | loading fbdialog data: /private/home/roller/working/parlai/data/ConvAI2/valid_self_original.txt
18:21:19 | 35.6% complete (2,775 / 7,801), 0:00:10 elapsed, 0:00:18 eta
    accuracy  exs  f1  interdistinct-1  interdistinct-2  intradistinct-1  intradistinct-2
           1 2775   1            .1091            .5404            .9576            .9968
18:21:29 | 52.5% complete (4,094 / 7,801), 0:00:20 elapsed, 0:00:18 eta
    accuracy  exs  f1  interdistinct-1  interdistinct-2  intradistinct-1  intradistinct-2
           1 4094   1           .08957            .4973            .9586            .9967
18:21:39 | 66.6% complete (5,194 / 7,801), 0:00:30 elapsed, 0:00:15 eta
    accuracy  exs  f1  interdistinct-1  interdistinct-2  intradistinct-1  intradistinct-2
           1 5194   1           .07928            .4711            .9587            .9968
18:21:49 | 78.6% complete (6,134 / 7,801), 0:00:40 elapsed, 0:00:11 eta
    accuracy  exs  f1  interdistinct-1  interdistinct-2  intradistinct-1  intradistinct-2
           1 6134   1           .07264            .4532            .9586            .9967
18:21:59 | 89.6% complete (6,986 / 7,801), 0:00:50 elapsed, 0:00:06 eta
    accuracy  exs  f1  interdistinct-1  interdistinct-2  intradistinct-1  intradistinct-2
           1 6986   1           .06808            .4400            .9582            .9968
18:22:09 | 99.5% complete (7,760 / 7,801), 0:01:00 elapsed, 0:00:00 eta
    accuracy  exs  f1  interdistinct-1  interdistinct-2  intradistinct-1  intradistinct-2
           1 7760   1           .06439            .4285            .9586            .9968
18:22:09 | Finished evaluating tasks ['convai2'] using datatype valid
    accuracy  exs  f1  interdistinct-1  interdistinct-2  intradistinct-1  intradistinct-2
           1 7801   1           .06425            .4280            .9587            .9968
parlai em -t convai2 --model repeat_label --metrics distinct  65.09s user 3.51s system 107% cpu 1:04.06 total
```

### CODE REVIEW

File: parlai/core/metrics.py [hunk 1]
```diff
@@ -523,6 +529,60 @@ def compute_many(
         )
 
 
+class IntraDistinctMetric(AverageMetric):
+    """
+    Compute intra-distinct (per-utterance).
+    """
+
+    @classmethod
+    def _ngram(cls, seq, n: int):
```
Comment (jxmsML) [inline parlai/core/metrics.py:538]: this looks like this NGramIterator for unlikelihood training: https://github.com/fairinternal/ParlAI-Internal/blob/master/projects/meena/safety/unlikelihood/agents.py#L33-L53 maybe we can extract those a separable feature?

Comment (stephenroller) [reply parlai/core/metrics.py:538]: I debated. I even have it copy pasted here. OTOH it's two lines of code lol

File: parlai/core/metrics.py [hunk 2]
```diff
@@ -523,6 +529,60 @@ def compute_many(
...
+        for i in range(len(seq) - n + 1):
```
Comment (jxmsML) [inline parlai/core/metrics.py:539]: as discussed, I'm a bit less certain about the scenario when `len(seq) < n`. Others LGTM

Comment (stephenroller) [reply parlai/core/metrics.py:539]: Manually checked and this sequence doesn't give partial n-grams. I think this implementation is faithful to the original.

File: tests/test_metrics.py [hunk 3]
```diff
@@ -385,5 +387,30 @@ def test_classifier_metrics(self):
...
+        m1 = InterDistinctMetric.compute("this test", 1)
+        self.assertAlmostEqual(m1, 1.0, delta=0.001)
+        m2 = InterDistinctMetric.compute("another test", 1)
+        self.assertAlmostEqual(m2, 1.0, delta=0.001)
+        # we now have 4 tokens, 3 words
+        self.assertAlmostEqual(m1 + m2, 3 / 4)
+
+    def test_intra_distinct(self):
+        # 4/5 are unique
+        m1 = IntraDistinctMetric.compute("this is some test test", 1)
```
Comment (jxmsML) [inline tests/test_metrics.py:408]: Do you think it makes sense to iteratively compute this metrics on multiple examples (averaged over multiple ones)

Comment (stephenroller) [reply tests/test_metrics.py:408]: I mean I do... I have my base case and my inductive case

Comment (jxmsML) [reply tests/test_metrics.py:408]: oh I see you have such tests for L404 and 412. No more questions from me.

**Review by jxmsML — APPROVED** [review]

**Pull request merged**